### DISCUSSION TITLE
sig_float

### DISCUSSION THREAD

**Discussion by steph-lheureux** [steph1111/PRECISE-Precision-Rules-Enforced-Calculations-In-Scientific-Environments#18] [General]
I think it works! What tests should we try? The ones that I came up with are passing

**Comment by schrodinstein** [comment]
The tests you have in 'create tests' look great. What's the easiest way to pass a ton of cases?

**Comment by steph-lheureux** [comment]
Honestly I am not sure how to run a ton of tests at once with this. I created those one at at time just stealing problems from worksheets I found online

**Comment by steph-lheureux** [comment]
I think I'll add a few more with multiple operations. But so far the ones I've added are passing!

**Comment by schrodinstein** [comment]
I added a quick script in my fork that can generate tests. Once I know how to import/call your data type, we can automate testing for tons of cases and then evaluate by hand.

**Comment by steph-lheureux** [comment]
Sounds great thanks! I'll check those out tomorrow

**Comment by steph-lheureux** [comment]
I started the README thing, take a look and let me know if this is what you were thinking. Note there is one broken link in the table of contents (multiplication), I am still working on that

**Comment by steph-lheureux** [comment]
Thanks for moving stuff around! When I started this repo I thought it would contain a bunch of different modules but now I think its best to have one repo per project/module. I am unsure how to move stuff around without breaking everything, could you also move the sig_float.py and tests_sig_float.py out of the significant figures folder and to the main one? Then I can rename the repo sig_float
No idea if this makes sense, some things are easier to explain in person sometimes

**Reply by schrodinstein** [reply]
Actually, I like your approach better. We can create a root-level readme that links the the module readmes.

**Comment by steph-lheureux** [comment]
I have units basically working!! But before I go too crazy with this idea how does this basic implementation seem? 
The numerator and denominator (if applicable) are assigned a list of units upon creation of the sig_float. Units with powers can be denoted using tuples.
 
```python
num1 = sig_float("2.8", ["kg", "m"], [("s", 2)])
num2 = sig_float("2.2", ["kg", "m"], [("s", 2)])
print(f"({num1}) + ({num2}) = {num1 + num2}")
```
```
(2.8 kg m/s^2) + (2.2 kg m/s^2) = 5.0 kg m/s^2
```

**Reply by steph-lheureux** [reply]
**Update:** I am thinking of changing this to a dictionary implementation. It makes more sense for coding efficiency

**Reply by steph-lheureux** [reply]
I apologize for the spam but units work and cancel with each other!

From the gas law video 
```python
V1 = sig_float("2.00", {"L":1})
P1 = sig_float("752.0", {"mmHg":1})
T1 = sig_float("293", {"K":1})
P2 = sig_float("2943", {"mmHg":1})
T2 = sig_float("273", {"K":1})
V2 = (T2 * P1 * V1) / (P2 * T1)
print(V2)
```
```
0.476 L
```

An example from my physics homework that involves lots of unit canceling
```python 
F1 = sig_float("28.4", {"kg":1, "m":1}, {"s":2})
g = sig_float("9.80", {"m":1}, {"s":2})
F2 = sig_float("17.0", {"kg":1, "m":1}, {"s":2})
density_fluid = sig_float("1000.", {"kg":1}, {"m":3})

density_obj = (F1 / g) / ((F1 - F2)/(density_fluid * g))
print(density_obj)
```
```
2490 kg/m^3
```

**Comment by steph-lheureux** [comment]
I have a function for LaTeX now but could you maybe give me an example of how you would want something formatted? I have 
`2490\frac{kg}{m^3}` which results in
![opt1](https://user-images.githubusercontent.com/96219204/235600349-c1792b45-f2bd-45e0-84bd-51e9fb9d9110.png)

Or the original output of `__str__()` (aka just print()) is `2490 kg/m^3` which makes this when put through a formatter
![opt2](https://user-images.githubusercontent.com/96219204/235600455-ed515052-19f2-4c76-aa88-ae9c309f7b85.png)
Or were you thinking of something completely different? If you can give me an example of how one number should be formatted with units in LaTeX I can write a function that handles all that

*Note: youre gonna have to switch to lightmode to see the inserted images*

**Reply by schrodinstein** [reply]
For single answers, what you have is fine. Maybe consider adding a space between the number and the units, or a multiplication dot: 

a · b

_Edit: We may also want to consider units on one line: L · atm · K⁻¹ · mol⁻¹_

_Or in the case of your example: 2490 kg · m⁻³_

Eventually, I would love to be able to show multiple steps with pretty formatting and cancelling of units:

<img width="940" alt="8 3-10" src="https://user-images.githubusercontent.com/17320307/235819749-cb24e22f-02b9-4201-b3da-1a8603e5a969.png">

**Reply by schrodinstein** [reply]
NOTE: Pretty formatting and cancelling are enhancements.

**Reply by steph-lheureux** [reply]
Awesome thank you so much! I'll add those "multiplication dots". 
I can definitely get units on one line if you would prefer that. This could even be an option in addition to what we have now, like the "mode" of output. Like how print has a sep option, `print("text", sep="-")` we could have a unit format option.  `num.latex(format=1)`  or `num.latex(format=2)` for one or two lines of units respectively. If so, what should be the default given no arguments, one or two lines?

**Reply by steph-lheureux** [reply]
Theres still issues but...
```python 
R = sig_float("0.08206", {"L":1, "atm":1, "mol":-1, "K":-1})
print(R.latex(format=1))
```
```
0.08206 \; L \cdot atm \cdot mol^{-1} \cdot K^{-1}
```

**Comment by steph-lheureux** [comment]
I am back working on this again, finals are doneee!! 

### New feature
I added a new feature for exact numbers that will not affect the sig figs/precision of the resultant. Upon declaration of a sig_float you can note that the number is exact. 
```python 
num1 = sig_float("1", exact=True)
```
</br>

### Issue resolved (#30)
The LaTeX()  function works again which is very exciting, the [code](https://github.com/steph1111/PRECISE/blob/79305fd6a3b2b8050c64388f2f98217af0cb98ba/sig_float/sig_float.py#L164) is special but it works well with no issues.
```python
R = sig_float("0.08206", {"L":1, "atm":1, "mol":-1, "K":-1}, exact=True)
print(R.latex(format=1))
print(R.latex(format=2))
print(R.latex(format=3))
```
```
0.08206 \; L \cdot atm \cdot mol^{-1} \cdot K^{-1}
0.08206 \frac{L \cdot atm}{mol \cdot K}
0.08206 \; L \cdot atm / mol \cdot K
```
$\large 0.08206 \\; L \cdot atm \cdot mol^{-1} \cdot K^{-1}$
$\large 0.08206 \frac{L \cdot atm}{mol \cdot K}$
$\large 0.08206 \\; L \cdot atm / mol \cdot K$
*Note: For some reason github's LaTeX "translator" is not recognizing the \; as a space, but my other editor does, \; is a space right??*
</br>

### Issue resolved (#33)
I rewrote the `sig_figs()` function to be much shorter, I am very happy with the [new implication](https://github.com/steph1111/PRECISE/blob/79305fd6a3b2b8050c64388f2f98217af0cb98ba/sig_float/sig_float.py#L107)
</br>

### Issue resolved (#36)
Implementation of sig_float raised to a int/float power using ** 
<br>

### What next?
I am working on resolving an issue with overlined zeros, but beyond that what should I work on next?

**Reply by schrodinstein** [reply]
Yes, in LaTeX, a \\; is recognized as a space. There are several different spaces, but I cannot recall them all at the moment.

**Reply by schrodinstein** [reply]
Ok, so I edited your comment to test a theory. It seems markdown requires double backslash "\\;" to work properly. LaTeX only uses a single.

**Reply by steph-lheureux** [reply]
Ohh okay cool good to know! I'll leave it as `\;` then

**Comment by schrodinstein** [comment]
To do for now:
Let's try making some simple problem set templates! Combined gas law? Density? Maybe 20 randomly generated problems in a given set. Then we can start looking into what needs to be done as far as solutions go.

Dimensional analysis? With LaTeX?? 

Enhancements:
I'd like to have working scientific notation at some point. Perhaps implement an option for SI prefixes? These could be parts of a separate piece of code. So output of 1000000 m from sig_float (w/2SF) could become 1.0e-9 m --or-- be 1.0 nm. If a number falls within a predetermined number of powers of 10 in either direction of a given prefix, it can be set to that prefix. i.e. 1-999 is m but 1000+ is km, etc.

**Reply by schrodinstein** [reply]
The could also be useful for problems that give units in mL but want L, etc.

**Reply by steph-lheureux** [reply]
Sounds good! I'll start working on that next. I'll see if python has anything nice for scientific notation. Also issue #31 is causing me some trouble, let me know if you have any ideas there

**Comment by steph-lheureux** [comment]
Im still working on this but have a robotics comp that has been taking up most of my time recently, I will be back on the project next week when I am back!

**Comment by steph-lheureux** [comment]
### General updates
The robotics competition I mentioned above is over so I am back to working on the code now. And exciting news, Cabrillo Robotics took [first place](https://20693798.fs1.hubspotusercontent-na1.net/hubfs/20693798/Winners%20-%202023%20Awards-MATE%20ROV%20Competition.pdf) in internationals!!

### Code updates 
I think I may have figured out outputting in scientific notation. I made this a feature of the latex() function, it seems to work but more tests are definitely needed 

### Question
I want to implement the option to provide the the number to the constructor in scientific notation. To do such, we need to decide on a standardized format for this. Would that be something like: `9.732e-06` or would you like it to be` 9.732x10^-06` or something else?

**Reply by steph-lheureux** [reply]
I am going to go with the standard way of `9.732e-06` for now but that should be easy to change in the future if need be

**Reply by steph-lheureux** [reply]
Exciting news, scientific works!! (Well thats as far as I can tell so far). You can now output in scientific (with latex) and input initial values in scientific

---
```python 
num1 = sig_float("120000", {"kg":1, "m":1, "s":-2})
print(num1.latex(sci=True)") 
```
$\large 1.2 \times 10^{5}  kg \cdot m \cdot s^{-2}$

---
```python
print(C.sig_figs()) 
print(C)
print(C.latex(sci=True))
```
```
3
30\bar{0}000000 \; m \cdot s^{-1}
3.00 \times 10^{8} \; m \cdot s^{-1}
```
$\large 30\bar{0}000000  m \cdot s^{-1}$
$\large 3.00 \times 10^{8}  m \cdot s^{-1}$
